collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1476,99 +1467,77 @@ public boolean dispatchTouchEvent(MotionEvent mv) {
-        } catch (JavaRosaException e) {
-            Timber.d(e);
-            createErrorDialog(e.getCause().getMessage(), DO_NOT_EXIT);
-        }
+    @Override
+    public void onSwipeBackward() {
+        moveScreen(BACKWARDS);
     }
 
-    /**
-     * If moving backwards is allowed, displays the view for the previous question or field list.
-     * Steps the global {@link FormController} to the previous question and saves answers to the
-     * data model without checking constraints.
-     */
-    public void showPreviousView() {
-        if (allowMovingBackwards) {
-            state = null;
-            FormController formController = getFormController();
-            if (formController != null) {
-                // The answer is saved on a back swipe, but question constraints are ignored.
-                if (formController.currentPromptIsQuestion()) {
-                    saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
-                }
+    private void moveScreen(Direction direction) {
+        FormController formController = getFormController();
+        if (formController == null) {
+            Timber.w("FormController has a null value");
+            swipeHandler.setBeenSwiped(false);
+            return;
+        }
 
-                try {
-                    if (formController.getEvent() != FormEntryController.EVENT_BEGINNING_OF_FORM) {
-                        int event = formController.stepToPreviousScreenEvent();
+        if (audioRecorderViewModel.isRecording().getValue()) {
+            // We want the user to stop recording before changing screens
+            DialogUtils.showIfNotShowing(RecordingWarningDialogFragment.class, getSupportFragmentManager());
+            swipeHandler.setBeenSwiped(false);
+            return;
+        }
 
-                        // If we are the beginning of the form, there is no previous view to show
-                        if (event == FormEntryController.EVENT_BEGINNING_OF_FORM) {
-                            event = formController.stepToNextScreenEvent();
-                            swipeHandler.setBeenSwiped(false);
+        if (direction == FORWARDS && formController.getEvent() == FormEntryController.EVENT_END_OF_FORM) {
+            // We're tryin to move forwards at the end of the form so just cancel
+            swipeHandler.setBeenSwiped(false);
+            return;
+        } else if (direction == BACKWARDS && !allowMovingBackwards) {
+            // We're not allowed to move backwards but trying to so just cancel
+            swipeHandler.setBeenSwiped(false);
+            return;
+        }
 
-                            if (event != EVENT_PROMPT_NEW_REPEAT) {
-                                // Returning here prevents the same view sliding when user is on the first screen
-                                return;
-                            }
-                        }
+        state = null; // This is needed for something but is definitely suspect
```

Comment:
Looks like it's used to keep track of which questions' guidance hints are expanded and the search text for selects with the search appearance. Would be great to move all that to a view model, I think.